js/ui/searchbar.js

```diff
@@ -232,14 +244,54 @@ define([
       }
     },
 
+    searchHistory: function() {
+      var input = document.find(".searchbar").find(".search-box");
+      var history = [];
+      var currentIndex = 0;
+      var temporaryQuery = '';
+
+      return {
+        add: function(query) {
+          if (query != history[history.length - 1]) { // deduplicate
+            history.push(query);
+          }
+          currentIndex = history.length;
+          temporaryQuery = '';
+        },
+
+        showPrevious: function() {
+          if (currentIndex == history.length) {
+            temporaryQuery = input.value;
+            // skip previous search if we're already showing the same value
+            if (temporaryQuery == history[currentIndex-1]) {
+              currentIndex--;
+            }
+          }
+          if (currentIndex - 1 >= 0) {
+            input.value = history[--currentIndex];
+          }
+        },
+
+        showNext: function() {
+          if (currentIndex + 1 < history.length) {
+            input.value = history[++currentIndex];
+          } else if (currentIndex + 1 == history.length) {
+            currentIndex++;
+            input.value = temporaryQuery;
+          }
+        }
+      }
+    }(),
```

Comment:
I realized that the way I'm adding this, searchHistory would only be created once and attached to the prototype, instead of with each construction of SearchBar. It doesn't break anything right now but it could down the road. I'll change the way this is done tonight so it's not like that. Hold on merging, but feel free to try it out.